File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleops/tuning/drive/ManualFeedbackTuner.java
package org.firstinspires.ftc.teamcode.teleops.tuning.drive;

import static org.firstinspires.ftc.teamcode.FieldConfig.tileSize;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.roadrunner.Pose2d;
import com.qualcomm.robotcore.eventloop.opmode.OpModeManager;

import org.firstinspires.ftc.teamcode.subsystems.VoltageSensor;
import org.firstinspires.ftc.teamcode.subsystems.localization.OdometryLocalizer;
import org.firstinspires.ftc.teamcode.teleops.ActionOpMode;
import org.firstinspires.ftc.teamcode.subsystems.MecanumDrive;
import org.firstinspires.ftc.teamcode.teleops.OpModeRegister;

@Config
public final class ManualFeedbackTuner extends ActionOpMode {
    public static double DISTANCE = 64;
    public static Pose2d startingPose = new Pose2d(2.0*tileSize, 2.0*tileSize, 0);


    @Override
    public void runOpMode() throws InterruptedException {
        if (OpModeRegister.DRIVE_CLASS.equals(MecanumDrive.class)) {
            MecanumDrive drive = new MecanumDrive(hardwareMap, startingPose, new OdometryLocalizer(hardwareMap), new VoltageSensor(hardwareMap));

            waitForStart();

            while (opModeIsActive()) {
                drive.getVoltageSensor().periodic();
                drive.periodic();

                Pose2d endPose = startingPose.times(new Pose2d(DISTANCE, 0, 0));
                runBlocking(
                        drive.actionBuilder(drive.getPose())
                                .lineToX(endPose.trans.x)
                                .lineToX(startingPose.trans.x)
                                .build());
            }
        }  else {
            throw new AssertionError();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleops/tuning/drive/ManualFeedbackTuner.kt
package org.firstinspires.ftc.teamcode.teleops.tuning.drive

import com.acmerobotics.dashboard.canvas.Canvas
import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.acmerobotics.roadrunner.Action
import com.acmerobotics.roadrunner.Pose2d
import org.firstinspires.ftc.teamcode.FieldConfig.tileSize
import org.firstinspires.ftc.teamcode.subsystems.MecanumDrive
import org.firstinspires.ftc.teamcode.subsystems.VoltageSensor
import org.firstinspires.ftc.teamcode.subsystems.localization.OdometryLocalizer
import org.firstinspires.ftc.teamcode.teleops.OpModeRegister
import org.firstinspires.ftc.teamcode.util.Action.ActionOpMode
import org.firstinspires.ftc.teamcode.util.Action.ParallelDeadlineAction

@Config
class ManualFeedbackTuner : ActionOpMode() {

    override fun runOpMode() {
        if (OpModeRegister.DRIVE_CLASS == MecanumDrive::class.java) {
            val drive = MecanumDrive(
                hardwareMap,
                startingPose,
                OdometryLocalizer(hardwareMap),
                VoltageSensor(hardwareMap)
            )
            waitForStart()
            while (opModeIsActive()) {
                runBlocking(ParallelDeadlineAction(
                    drive.actionBuilder(drive.pose)
                        .lineToX(startingPose.trans.x + DISTANCE)
                        .lineToX(startingPose.trans.x)
                        .build(),
                    object : Action {
                        override fun run(telemetryPacket: TelemetryPacket): Boolean {
                            drive.voltageSensor.periodic()
                            telemetryPacket.put("voltage", drive.voltageSensor.voltage)
                            return true
                        }

                        override fun preview(canvas: Canvas) {}
                    }
                ))
            }
        } else {
            throw AssertionError()
        }
    }

    companion object {
        var DISTANCE = 64.0
        var startingPose = Pose2d(-2.0 * tileSize, 2.0 * tileSize, 0.0)
    }
}